[WIP] EZP-25946: Procedure to migrate existing images/binaries to another IO adapter

glye opened the pull request:
> Implements https://jira.ez.no/browse/EZP-25946
> Ready for review

Migrates files from local storage to NFS/DFS cluster, and vice versa. (Theoretically, any Flysystem adapter should work.)

Adds a Command which validates from/to input and runs the migration in batches. Input options partially based on [legacy clusterize.php](https://github.com/ezsystems/ezpublish-legacy/blob/d76ce9824dabaceeefa77530c8c611e447d1b109/bin/php/clusterize.php) but made more generic. Meaning it's not just for clusterize/unclusterize, but also for migrating between e.g. NFS and AWS/S3.

`$ php app/console ezplatform:io:migrate-files --dry-run --from=default,default --to=dfs,nfs`

- [x] ~~[We can't differentiate between media and binary files](https://github.com/ezsystems/ezpublish-kernel/blob/7170576929fefd6c1bb240f13553b14ca721ca43/eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php#L179), as they have the same prefix ("original"). Problem?~~ Migrate all in one go.
- [x] ~~Migration handlers (binary, media, image). Do we actually need 3 of them? Only purpose now is to allow skipping some of them, and even that is unreliable (see previous point)~~ Use only one.
- [x] File handling is normally one-by-one. For migration we need to iterate over *all* files, and ideally count them first. Need support for that in handlers.
- [x] ~~Consider using the [Replicate Adapter](http://flysystem.thephpleague.com/adapter/replicate/) for Flysystem-Flysystem migration~~ Rejected for now. Kind of a different use-case.
- [x] Logging
- [x] Some adapters may not support counting - if so we have to detect that and work around it.
- [x] ~~Since FlySystem doesn't support fetches with limit/offset we are currently loading metadata for *all* files - likely a memory/performance problem for big data.~~ Using DB iterators now.
- [x] ~~Implement source files deletion (or remove the `--remove-files` option)~~ Option removed.
- [x] ~~Doc `IOMetadataHandler`/`IOBinarydataHandler` additions~~ They're gone now.
- [x] Make classes `final` to avoid future BC issues on this, until we're ready
- [ ] Tests (including possibly an integration test doing a full migration)

glye commented:
@bdunogier Early review ping, or we'll talk tomorrow.

glye commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,276 @@
+
+                $progress->clear();
+
+                $output->write("\r");
```
This is unfinished stuff. It will become obvious with larger datasets that this won't do, update coming soon. (It turns out that the ProgressBar supports all we need, including running without knowing the total count.)

glye replied:
Rewritten now to support uncountable filesystem adapters.

nicolas-bastien commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,276 @@
+        }
+
+        $output->writeln([
+            "Migrating from '$fromHandlers[0],$fromHandlers[1]' to '$toHandlers[0],$toHandlers[1]'",
```
If you don't wanna bother about quote you can do : 

 sprintf("Migrating from '%s' to '%s'", implode(',', $fromHandlers), implode(',', $toHandlers));

this is a useful trick when arrays are longer.

glye replied:
Yes, but IMHO this is less readable in this case.

bdunogier replied:
Oh, no... one of _those_ discussions.

glye replied:
If both reviewers agree I'll change it - and bury the bitter resentment deep in my black old heart :)

nicolas-bastien commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,276 @@
+
+        $fromHandlers = $input->getOption('from') ? explode(',', $input->getOption('from')) : null;
+        $toHandlers = $input->getOption('to') ? explode(',', $input->getOption('to')) : null;
+        if (!$this->areHandlerOptionsValid($fromHandlers, $toHandlers, $output)) {
```
in a more sf way I would name this : valiateHandlerOptions

Validator are made like that and you don't have to bother about is or are

bdunogier replied:
Also, since they do the same thing, I'd use a method for each, and call it two times.

glye replied:
@nicolas-bastien Renamed, thanks.
@bdunogier It's running in loops now: More compact, no duplication, less readable :)

nicolas-bastien commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,276 @@
+            }
+        }
+
+        if ($fromHandlers && $toHandlers && $fromHandlers === $toHandlers) {
```
here I guess $fromHandlers === $toHandlers is enough ?

glye replied:
No, because I support @bdunogier's proposal that if you don't specify from/to they will be set to default handler and first configured handler, respectively. However if I do validation after that point, I could reduce the if like this indeed. And that's probably a good idea, because you could potentially configure your site so that the default from/to would end up being the same.

bdunogier replied:
@glye maybe the default could be set before validating, in order to limit coupling ? If from/to aren't set, set them to the defaults.

glye replied:
@bdunogier Too late, I have already suggested that myself :) Done now.

nicolas-bastien commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,276 @@
+     * @return bool
+     */
+    protected function areHandlerOptionsValid(
+        $fromHandlers,
```
As you are doing the same validation for fromHandlers and then toHandlers I guess you can to a single method and call it two times

glye replied:
Then I can't validate that the two handlers are different from each other. But I can do the first part of the validation in a 2-step loop instead.

bdunogier replied:
Good point @glye ! An alternative could be to make it `validateHandlerOptions(InputInterface $input, OutputInterface $output)`. Maybe with a loop inside to avoid code duplications.

glye replied:
It's based on loops now. I'm still passing the handler arrays, since now I'm setting defaults before validating.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,276 @@
+
+        $helper = $this->getHelper('question');
+        $question = new ConfirmationQuestion(
+            '<question>Are you sure you want to proceed?</question> ',
```
Wide topic: do we want to make those translatable ? :)

glye replied:
No, because I have do lots of rewriting then :) AFAIK we have never translated console scripts before. We can change that of course, but it's indeed a wide topic.

glye commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,276 @@
+
+class MigrateFilesCommand extends Command
+{
+    private $configuredMetadataHandlers;
```
Added, thanks.

glye commented:
Thanks reviewers! Updated accordingly, pretty much. Other changes:
- Validate the `--bulk-count` option
- Support uncountable filesystem adapters
- Implemented progressbar display properly
- Added magic to ensure the progressbar updates ca. once per second, regardless of migration performance. If the magic isn't popular, we can make it update for every N file instead, starting with 10 and see how that goes with larger datasets.

glye commented:
Added some tests, more to come.

glye commented:
> The methods that lists & counts files could be moved to their own interface. Then we may realize that we could do what legacy does, and use the SPI to list files (legacy migrates the stuff that's referenced in `ezimagefile` and `ezbinaryfile`).

@bdunogier Not sure I follow. Won't that only work in one direction?

glye commented:
> what about separating listing & migrating files?
> Then we may realize that we could do what legacy does, and use the SPI to list files

Blatant begging-the-question-ism aside :P that could be good. Like a `MetadataMigrationHandler` and a `BinarydataMigrationHandler` then. I don't know if the SPI has functionality for fetching all file info in batches either, but I see the point that it can be safer/better to implement it there, avoiding Flysystem limitations.

glye commented:
Update: Separated file listing from file migration. Implemented db-based listing of images. Binary and media are todo.

glye commented:
Update: Completed separation of file listing from file migration, for images, binary and media. Image listing uses iterator that already exists related to image variation purging. A new iterator is added for binary and media. Existing api/spi is no longer touched, apart from bug fix https://github.com/ezsystems/ezpublish-kernel/pull/1825/commits/9b12bb7a3fa32d1228badf65c4194d7426785de5

bdunogier commented:
I'm thinking it might be worth sending 9b12bb7 to its own PR. Looks interesting :)

What do you think GL ?

glye commented:
> I'm thinking it might be worth sending 9b12bb7 to its own PR.

After the rewrite I no longer depend on it, so that's unproblematic.

glye commented:
Update: Marked as experimental in doc and in the command itself. Made classes/methods `final` to avoid BC issues until fully finished.

glye commented on doc/specifications/io/io_migration_script.md:
```diff
@@ -0,0 +1,74 @@
+A common use-case is to migrate local files, stored with the default IO
+configuration, to a new IO configuration.
+
+### 
```
Thanks, removed as no text needed.

glye commented:
Update: When total file count is known, and some files could not be migrated (the file for the given db entry does not exist) show the count of files that failed.

Tested on the Studio demo site with 260 images. Works fine for main images. Does not migrate variations from `_aliases` directory, meaning the variation path generator fails. This failure is not directly visible since the variations are generated on demand anyway, but it harms performance.

glye commented:
Update: Fixed `_alias` migration by using the correct service alias: `@ezpublish.image_alias.variation_path_generator`

glye commented:
@bdunogier @andrerom @nicolas-bastien Ready for review now, nothing further on the horison other than tests.

Is this usage of `ezsettings.default` ok? If not, what's the right way to get the non-default, overridden setting? https://github.com/ezsystems/ezpublish-kernel/pull/1825/files#diff-63027c1579de5d9ddeda5db593953b7cR42

andrerom commented on doc/specifications/io/io_migration_script.md:
```diff
@@ -0,0 +1,74 @@
+# IO migration
+
+> Added in 6.7 / 2016.12
```
6.8, and you can drop the year syntx, versions are following semantic versioning now _(besides the yearly LTS, aka 2017LTS marketing name)_.

glye replied:
Thanks, done.

bdunogier commented:
I'll review this today.

andrerom commented:
What about today? 😃😛

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileListerRegistry/IdentifierBased.php:
```diff
@@ -0,0 +1,57 @@
+/**
+ * A registry of FileListerInterfaces that uses an identifier string to identify the lister.
+ */
+final class IdentifierBased implements FileListerRegistry
```
The naming could be improved here, in regards to interface/implementation: the interface _is_ already about identifiers, since it has `getItem($identifier)` and `getIdentifiers()` methods.

It could be a ConfigurableRegistry (as you configure it by means of its constructor arguments), or maybe ArrayBasedRegistry, as it uses an array internally.

What do you think ?

glye replied:
I think this gives me a headache :) ConfigurableRegistry works if I remove the constructor from the interface (which I should, and just did). I'd say Configurable is a more important aspect than the internal array, so I'll go for that.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileListerRegistry/IdentifierBased.php:
```diff
@@ -0,0 +1,57 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileListerRegistry.php:
```diff
@@ -0,0 +1,38 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileListerInterface.php:
```diff
@@ -0,0 +1,31 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileLister/ImageFileLister.php:
```diff
@@ -0,0 +1,97 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileLister/FileRowReaderInterface.php:
```diff
@@ -0,0 +1,36 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileLister/FileRowReader/LegacyStorageMediaFileRowReader.php:
```diff
@@ -0,0 +1,22 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileLister/FileRowReader/LegacyStorageFileRowReader.php:
```diff
@@ -0,0 +1,69 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileLister/FileRowReader/LegacyStorageBinaryFileRowReader.php:
```diff
@@ -0,0 +1,22 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileLister/FileIteratorInterface.php:
```diff
@@ -0,0 +1,19 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileLister/FileIterator/LegacyStorageFileIterator.php:
```diff
@@ -0,0 +1,80 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileLister/BinaryFileLister.php:
```diff
@@ -0,0 +1,71 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/EzPublishIOBundle.php:
```diff
@@ -28,6 +28,9 @@ public function build(ContainerBuilder $container)
                 $extension->getBinarydataHandlerFactories()
             )
         );
+        $container->addCompilerPass(
```
Nitpick: no need to wrap this one. The one above was wrapped because it had long arguments.

glye replied:
Fixed.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/DependencyInjection/Compiler/MigrationFileListerPass.php:
```diff
@@ -0,0 +1,40 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,312 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/MigrationHandlerInterface.php:
```diff
@@ -0,0 +1,30 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/MigrationHandler.php:
```diff
@@ -0,0 +1,79 @@
+<?php
+
```
Missing copyright block.

glye replied:
@bdunogier Uh, no, it's right below on line 6. Commented on the wrong file, maybe? 😕

bdunogier replied:
Alcohol is a bane :)

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileMigratorInterface.php:
```diff
@@ -0,0 +1,23 @@
+<?php
+
```
Missing copyright block.

bdunogier commented on eZ/Bundle/EzPublishIOBundle/Migration/FileMigrator/FileMigrator.php:
```diff
@@ -0,0 +1,59 @@
+<?php
+
```
Missing copyright block.

glye commented on eZ/Bundle/EzPublishIOBundle/Resources/config/io.yml:
```diff
@@ -1,8 +1,92 @@
+        parent: ezpublish.core.io.migration.migration_handler
+        arguments:
+            - "@ezpublish.core.io.migration.file_lister.file_iterator.binary_file_iterator"
+            - "%ezsettings.default.binary_dir%"
```
@bdunogier Yes I'd like to get the setting from the current siteaccess, but `$binary_dir$` is apparently not it, it gets interpreted as a literal string.

bdunogier replied:
Hmmm... it should really work :/

glye replied:
@bdunogier Yah. It works here: https://github.com/ezsystems/ezpublish-kernel/blob/e2d26faa54715675fc8f1eef3defbe192973bf5b/eZ/Bundle/EzPublishCoreBundle/Resources/config/fieldtype_services.yml#L232
when tested in web context. Is the problem that console scripts know nothing about site accesses? I tried specifying `--siteaccess`, no change (should use the default anyway). Now wondering if my Command needs to do something siteaccess/config related for this to work...

glye replied:
Actually the project container gets built wrong, it sets the resolved binary_dir as second param of the BinaryFileLister ctor, when it should be the fifth. The fifth param remains the litteral string. Weird.
```php
return $this->services['ezpublish.core.io.migration.file_lister.binary_file_lister'] =
    new \eZ\Bundle\EzPublishIOBundle\Migration\FileLister\BinaryFileLister(
        $this->get('ezpublish.core.io.metadata_handler.factory'),
        $this->get("ezpublish.config.resolver")->getParameter("binary_dir", null, null),
        $this->get('logger'),
        $this->get('ezpublish.core.io.migration.file_lister.file_iterator.binary_file_iterator'),
        '$binary_dir$');
```

glye commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,312 @@
+    ) {
+        $this->configuredMetadataHandlers = $configuredMetadataHandlers;
+        $this->configuredBinarydataHandlers = $configuredBinarydataHandlers;
+        $this->configuredMetadataHandlers['default'] = [];
```
@bdunogier They are silently ignored. I could ignore them more loudly (log warning or even fail).

glye replied:
I allowed them instead.

glye commented:
Update:
- `--list-io-configs` changed to `--list-io-handlers`
- Removed `File containing the ...` doc lines
- Privatised some `MigrationHandler` properties
- Other suggested improvements

glye commented on eZ/Bundle/EzPublishIOBundle/Migration/MigrationHandler.php:
```diff
@@ -0,0 +1,79 @@
+use eZ\Bundle\EzPublishIOBundle\ApiLoader\HandlerFactory;
+use Psr\Log\LoggerInterface;
+
+abstract class MigrationHandler implements MigrationHandlerInterface
```
Added.

glye commented on eZ/Bundle/EzPublishIOBundle/Migration/MigrationHandler.php:
```diff
@@ -0,0 +1,79 @@
+    protected $binarydataHandlerFactory;
+
+    /** @var \Psr\Log\LoggerInterface */
+    protected $logger;
```
Privatised those I could. For the remaining I'd have to refactor.

glye replied:
@bdunogier I could privatise the data handler properties and make final protected getters for them. OK?

glye commented on eZ/Bundle/EzPublishIOBundle/Migration/FileMigratorInterface.php:
```diff
@@ -0,0 +1,23 @@
+
+use eZ\Publish\SPI\IO\BinaryFile;
+
+interface FileMigratorInterface extends MigrationHandlerInterface
```
Added.

glye commented on doc/specifications/io/io_migration_script.md:
````diff
@@ -0,0 +1,74 @@
+```yaml
+ez_io:
+    binarydata_handlers:
+        // native binarydata handler, filesystem
````
Fixed.

glye commented on doc/specifications/io/io_migration_script.md:
````diff
@@ -0,0 +1,74 @@
+```
+
+### Console script
+Migrating is done by running the `ezplatform:io:migrate-files` console script:
````
Yeah, removed.

glye commented on doc/specifications/io/io_migration_script.md:
```diff
@@ -0,0 +1,74 @@
+Migration expects that IO handlers specified in the `from` and `to` 
+arguments are configured in the `ez_io` section.
+
+Without any argument, the command migrates files from the `default`
```
What happened to the joy of exploring uncharted territory? ;)
Added.

glye commented on doc/specifications/io/io_migration_script.md:
````diff
@@ -0,0 +1,74 @@
+
+Other script options are:
+```
+--list-io-configs          List available IO configurations
````
Yeah, changed to handlers.

glye commented on eZ/Bundle/EzPublishIOBundle/Command/MigrateFilesCommand.php:
```diff
@@ -0,0 +1,312 @@
+    private $fileMigrator;
+
+    public function __construct(
+        $configuredMetadataHandlers,
```
Added.

glye commented:
Update: Renamed the registry to `ConfigurableRegistry`. Allowed customising the `default` IO handlers.

andrerom commented on doc/specifications/io/io_migration_script.md:
```diff
@@ -0,0 +1,75 @@
+# IO migration
+
+> Added in 6.8
```
6.9 or most likely 6.10 by now

glye commented:
CS fix.

glye commented:
Rebased on master since https://github.com/ezsystems/ezpublish-kernel/pull/1856 is merged now. (I think I'll squash this and do the PR against 6.7 instead.)

glye commented:
Closing, continued in https://github.com/ezsystems/ezpublish-kernel/pull/1972
